cmd/podman/images/scp.go

```diff
@@ -0,0 +1,317 @@
+		return "", err
+	}
+	return "", nil
+}
+
+// parseLocation returns the valid connection data based off of the information provided by the user
+// iden is a string array of ssh identity keys, url is a string array of paths to ssh clients. and serv is
+// an array of service destinations
+func parseLocation(args []string, cfg *config.Config) ([]string, []string, map[string]config.Destination) {
+	iden := []string{}
+	url := []string{}
+	serv := map[string]config.Destination{}
+	if strings.Contains(args[0], "::") {
+		if len(args) == 2 { // if we have a connection name specified
+			if !(strings.Contains(args[1], "::")) { // if an image is specified, this mean we are loading from our client
+				scpOpts.Connections = append(scpOpts.Connections, args[0])
+				scpOpts.ToRemote = true
+			} else if strings.Contains(args[1], "::") { // both remote clients
+				scpOpts.FromRemote = true
+				scpOpts.ToRemote = true
+				if len(strings.Split(args[0], "::")) == 1 || (len(strings.Split(args[0], "::")) == 2 && len(strings.Split(args[1], "::")) == 2) { // is save->load w/ one image name or save-load w/ two image names
+					scpOpts.Connections = append(scpOpts.Connections, args[0])
+					scpOpts.Connections = append(scpOpts.Connections, args[1])
+				} else { // else its a load save (order of args)
+					scpOpts.Connections = append(scpOpts.Connections, args[1])
+					scpOpts.Connections = append(scpOpts.Connections, args[0])
+				}
+			}
+		} else { // else we are loading from the ssh/remote client
+			scpOpts.FromRemote = true
+			scpOpts.Connections = append(scpOpts.Connections, args[0])
+		}
+	} else if len(args) == 2 && strings.Contains(args[1], "::") && len(scpOpts.Connections) < 2 { // if we are given image host::
+		scpOpts.ToRemote = true
+		scpOpts.Connections = append(scpOpts.Connections, args[1])
+	}
+
+	for i, val := range scpOpts.Connections {
+		splitEnv := strings.SplitN(val, "::", 2)
+		switch len(splitEnv) {
+		case 1:
+			scpOpts.Connections[i] = splitEnv[0]
+		case 2:
+			if i == 0 {
+				scpOpts.ImageName = splitEnv[1]
+			} else if i == 1 {
+				scpOpts.Tag = splitEnv[1]
+			}
+			scpOpts.Connections[i] = splitEnv[0]
+		}
+
+		found := false
+		for name, data := range cfg.Engine.ServiceDestinations { // loop through all of the connections until we find ours
+			if name == scpOpts.Connections[i] {
+				iden = append(iden, data.Identity)
+				url = append(url, data.URI)
+				dst := config.Destination{}
+				dst.URI = url[0]
+				dst.Identity = iden[0]
+				serv[name] = dst
+				found = true
+				break
+			}
+		}
+		if !found { // no match, warn user and do a manual connection.
+			url = append(url, "ssh://"+scpOpts.Connections[i])
+			iden = append(iden, " ")
+			logrus.Warnf("Unknown connection name given. Please use system connection add to specify the default remote socket location")
+		}
+	}
+	if len(scpOpts.Connections) == 0 { // we default to remote or whatever the default connection is
+		connStr := cfg.Engine.ActiveService
```

Comment:
@jwhonce Overall it seems to me that a lot of this must already exist somewhere in (or distributed throughout?) Podman, around `pkg/bindings.sshClient`.

I haven’t looked into this deeply, and I’m not sure whether some of the probably special features of this subcommand (connecting to a hostname directly without having a system connection set up) are necessary or even desirable. (It seems likely that this was a deeply discussed design point already.)

Reply:
yes and no @jwhonce, I depend on the name of the connection for alot of the validation (remote, fedora etc) so I think keeping it as is might be the same since I will have to do an `indexOf` loop to find the name of the iden and uri that `ActiveService()` returns